irmin-pack.unix: Add an interface.

adatario opened the pull request:
This adds an interface file (irmin_pack_unix.mli) for the irmin-pack.unix package.

This is a major API change as it hides a lot of internal functionality that was previously exposed (the interface in `s.ml` was not being used in the interface of the package).

Many tests were using internal functionality. The required functionality has been exposed or the tests have been changed to use the public API.

This is a bit of a pandora's box but a good tour trough the code for me. Currently tests do not compile, I'm in the process of fixing them one-by-one. Once the tests compile and run I would have to check what clients are using.

I'm not sure if this is sensible or worth the effort, so your thoughts are appreciated.

metanivek commented on src/irmin-pack/unix/s.ml:
```diff
@@ -244,6 +244,24 @@ module type S = sig
     end
   end
+
+  (** {1 Internals} *)
```
I'm not sure how others feel, but I'm inclined to go further and wrap this part of the signature in an explicit `Internal` (or `Private`) module with the appropriate warning documentation like we do in https://github.com/mirage/irmin/blob/ae6b4fee5767b4c5e3c5c711f68e5b45281a740a/src/irmin-pack/inode_intf.ml#L125-L127

metanivek reviewed (commented):
I'm very much in favor of being explicit about our public interface. Thanks for digging into this!

Once you get all of our code passing, the only client code (to my knowledge -- perhaps double-check https://sherlocode.com) to evaluate will be tezos, which _shouldn't_ be an issue but will be interesting to see. :smile:

metanivek commented:
@icristescu could you shed some light on the difference between the `mem` functions? I'm not sure off the top of my head if they are functionally different.

icristescu commented on test/irmin-pack/test_flush_reload.ml:
```diff
@@ -135,7 +128,10 @@ let test_one t ~(ro_reload_at : phase_flush) =
           | `After_dict -> aux S2_after_flush_dict
           | `After_suffix -> aux S3_after_flush_suffix
         in
-        let () = Store.S.X.Repo.flush_with_hook ~hook rw in
```
since you are replacing the `*_with_hook` functions here, maybe remove them from store.ml too?

adatario replied:
Done with 6667f4f4e91668c14b218db528fe364ad5352369.

icristescu commented:
> I don't really understand what suffix_*_mem does as opposed to S.Backend.*.mem

The difference is that `S.Backend.*.mem` calls https://github.com/mirage/irmin/blob/main/src/irmin-pack/unix/pack_store.ml#L242 whereas your new functions call `unsafe_find` which:1/ does not return lwt and 2/can take the specialised `check_integrity` argument. It might be that they are not necessary in the test you are looking at and you could try to replace them with  `S.Backend.*.mem`. But if they are used inside a `hook` for example, you need to use a function which does not return lwt.

metanivek commented on src/irmin-pack/unix/store_intf.ml:
```diff
@@ -279,3 +309,29 @@ module type Maker_persistent =
+       and type Schema.Branch.t = string
+end
+
+(* This *_intf module does not have a Sigs module type as it is
+   directly exposed in the top-level interface
+   (irmin_pack_Unix.mli). *)
```
Let's follow the pattern and add a `store.mli` that uses the the `Sigs` signature (and probably adds a `Maker` module). Then `irmin_pack_unix.mli` can use `Store.S`, etc.

adatario commented:
> But if they are used inside a `hook` for example, you need to use a function which does not return lwt.

That seems to be it. They are used from a hook (https://github.com/mirage/irmin/blob/95bd8fd6a76c942e9f09fec9038d5bbb8132dc8a/test/irmin-pack/test_flush_reload.ml#L127-L129).

Thanks for your insight!

metanivek commented on doc/irmin-pack/index.mld:
```diff
@@ -2,6 +2,8 @@
 
 The [irmin-pack] package provides a space-optimized, on-disk backend inspired by {{:https://git-scm.com/book/en/v2/Git-Internals-Packfiles} Git Packfiles}.
 
+[irmin-pack] is desigend and developed for a very specific use-case and requires backend specific integration from applications. Users interested in a more general and stable Irmin backend are recommended to use other backend such as {!Irmin_fs} or {!Irmin_git}.
```
Let's remove the `and stable` part. :sweat_smile: As a back-end, it is quite stable (it has the heaviest production usage that I know of), but it is a bit more purpose-built in its intended usage and some of its API, as you point out.

adatario replied:
I meant stable in the sense of API stability. Of course as back-end it is stable :).

I've changed the note to following:

> Users interested in a more general Irmin backend with a more stable API are recommended to use other backend such as {!Irmin_fs} or {!Irmin_git}.

Is that ok?

I think the warning is fair. As a prospective Irmin user I would be very interested in API stability.

metanivek replied:
Aha! I see the distinction you're making now.

I agree with giving guidance to users, but also don't want to unnecessarily guide them away from `irmin-pack` since I do think it is probably the most performant and stable backend given its extensive real-world usage.

When I think of the API of `irmin-pack` and `irmin`, here is how I break it down:

- Configuration. This is the most complex and application-specific for `irmin-pack`, but helped by the examples (thanks for adding another!).
- Core Irmin API. This is the same across all backends so not applicable to the guidance.
- The extensions to the store API. This area is arguably less stable (I'm not sure how it has changed historically) but this PR makes it much better from an external integration standpoint (we won't accidentally leak all our internal APIs now!).

If the above corresponds to what you are thinking when you think about the API and its stability, what do you think the best guidance to give is?

To me, it is something like:
> [irmin-pack] is designed and developed for a specific use-case and requires more in-depth backend configuration from applications. It offers additional functionality through extensions to the core Irmin API, but users interested in a more general Irmin backend with simpler configuration are recommended to use other backend such as {!Irmin_fs} or {!Irmin_git}.
>
> See [examples/irmin-pack] for examples of how to configure and use some of the additional functionality.

adatario replied:
> If the above corresponds to what you are thinking when you think about the API and its stability, what do you think the best guidance to give is?

I tend to agree with your break down. I just feel that a big API, even if it not used, is a considerable complexity factor that should be considered by prospective users.

In any case, I like your suggested wording and have pushed 25b40c0e34b7e733215823b29e161859a2fc6300 that uses it.

Thanks!

metanivek commented on test/irmin-pack/test_flush_reload.ml:
```diff
@@ -23,20 +23,13 @@ let suffix_mem repo e =
-          (Store.S.X.Repo.contents_t repo)
-          k
-        |> Option.is_some
+        (* Store.S.X.Commit.CA.unsafe_find ~check_integrity:false *)
+        (*   (snd (Store.S.X.Repo.commit_t repo)) *)
+        (*   k *)
+        (* |> Option.is_some *)
```
These comments should be removed (I think?).

adatario replied:
Absolutely right. Removed them with 4632746b1f9ce94215cadba147ce78180b443630.

metanivek reviewed (changes requested):
A couple small changes needed before merging (and looks like you need to `dune build @fmt` since CI is failing for the fmt linter).

adatario commented:
> and looks like you need to `dune build @fmt` since CI is failing for the fmt linter

d'oh. Fixed (and squashed into edaf7aefdc577d67cd5d5a817d7300b4dd83cb8d).

metanivek commented:
Happens to the best of us! Thanks CI robot! :robot: 

It looks like there is a conflict that needs resolving with `main`, but I think this is ready once we finalize the language we want for giving users guidance. Thanks a lot for this PR. I'm super happy to improve our documentation and structure of APIs.

metanivek reviewed (approved):
Pending resolution of the conflict with `main` this LGTM! Thanks!

adatario commented:
> Pending resolution of the conflict with `main` this LGTM! Thanks!

Should be resolved and pushed!
